packages/server-gateway-interface/src/index.ts

```diff
@@ -0,0 +1,135 @@
+import type { KeyValueCache } from '@apollo/utils.keyvaluecache';
+import type {
+  DocumentNode,
+  ExecutionResult,
+  GraphQLError,
+  GraphQLFormattedError,
+  GraphQLSchema,
+  OperationDefinitionNode,
+} from 'graphql';
+import type { Logger } from '@apollo/utils.logger';
+import type { Trace } from '@apollo/usage-reporting-protobuf';
+
+export interface GatewayInterface {
+  onSchemaLoadOrUpdate(
+    callback: GatewaySchemaLoadOrUpdateCallback,
+  ): GatewayUnsubscriber;
+  load(options: { apollo: GatewayApolloConfig }): Promise<GatewayLoadResult>;
+  stop(): Promise<void>;
+}
+
+export type GatewaySchemaLoadOrUpdateCallback = (schemaContext: {
+  apiSchema: GraphQLSchema;
+  coreSupergraphSdl: string;
+}) => void;
+
+export type GatewayUnsubscriber = () => void;
+
+export interface GatewayApolloConfig {
+  key?: string;
+  keyHash?: string;
+  graphRef?: string;
+}
+
+export interface GatewayLoadResult {
+  executor: GatewayExecutor | null;
+}
+
+export type GatewayExecutor = (
+  requestContext: GatewayGraphQLRequestContext,
+) => Promise<GatewayExecutionResult>;
+
+export type GatewayExecutionResult = ExecutionResult<
+  Record<string, any>,
+  Record<string, any>
+>;
+
+// Note that the default value for TContext is the same as in AS3, not
+// BaseContext.
+export interface GatewayGraphQLRequestContext<TContext = Record<string, any>> {
+  readonly request: GatewayGraphQLRequest;
+  readonly response?: GatewayGraphQLResponse;
+  logger: Logger;
+  readonly schema: GraphQLSchema;
+  readonly schemaHash: GatewaySchemaHash;
+  readonly context: TContext;
+  readonly cache: KeyValueCache;
+  readonly queryHash: string;
+  readonly document: DocumentNode;
+  readonly source: string;
+  readonly operationName: string | null;
+  readonly operation: OperationDefinitionNode;
+  readonly errors?: ReadonlyArray<GraphQLError>;
+  readonly metrics: GatewayGraphQLRequestMetrics;
+  debug?: boolean;
+  readonly overallCachePolicy: GatewayCachePolicy;
+}
+
+export interface GatewayGraphQLRequest {
+  query?: string;
+  operationName?: string;
+  variables?: Record<string, any>;
+  extensions?: Record<string, any>;
+  http?: GatewayHTTPRequest;
+}
+
+export interface GatewayHTTPRequest {
+  readonly method: string;
+  readonly url: string;
+  readonly headers: GatewayHTTPHeaders;
+}
+
+export interface GatewayGraphQLResponse {
+  data?: Record<string, any> | null;
+  errors?: ReadonlyArray<GraphQLFormattedError>;
+  extensions?: Record<string, any>;
+  http?: GatewayHTTPResponse;
+}
+
+export interface GatewayHTTPResponse {
+  readonly headers: GatewayHTTPHeaders;
+  status?: number;
+}
+
+export interface GatewayHTTPHeaders {
```

Comment:
Can this just be a re-export of `FetcherHeaders`?

Reply:
Ooh, nice, I think you're right. FetcherHeaders also has `extends Iterable<[string, string]>` but that shouldn't hurt.

Reply:
(Request and Response are intentionally smaller though, but you didn't suggest that :) )